pkg/wait/wait_test.go

```diff
@@ -1,26 +1,338 @@
+	Context("ForDaemonSetReady", testForDaemonSetReady)
+	Context("ForPodsRunning", testForPodsRunning)
 })
 
+func testForPodsRunning() {
+	namespace := "test-namespace"
+
+	var (
+		initialPods   []*corev1.Pod
+		client        *testClient
+		waitCh        chan error
+		listError     error
+		numReplicas   int
+		labelSelector string
+	)
+
+	BeforeEach(func() {
+		listError = nil
+		initialPods = nil
+		numReplicas = 1
+		labelSelector = "app=test"
+	})
+
+	JustBeforeEach(func() {
+		var initObjs []runtime.Object
+		for _, p := range initialPods {
+			initObjs = append(initObjs, p)
+		}
+
+		client = &testClient{Client: fake.NewFakeClientWithScheme(scheme.Scheme, initObjs...), returnError: listError}
+
+		waitCh = runAsync(func() error {
+			return wait.ForPodsRunning("east", client, namespace, labelSelector, numReplicas)
+		})
+	})
+
+	newPod := func(name string, phase corev1.PodPhase) *corev1.Pod {
+		return &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      name,
+				Labels:    map[string]string{"app": "test"},
+			},
+			Status: corev1.PodStatus{
+				Phase: phase,
+			},
+		}
+	}
+
+	When("the Pods initially exist", func() {
+		BeforeEach(func() {
+			numReplicas = 2
+			initialPods = append(initialPods, newPod("test-Pod1", corev1.PodPending), newPod("test-Pod2", corev1.PodPending))
+		})
+
+		Context("and are initially running", func() {
+			BeforeEach(func() {
+				for _, p := range initialPods {
+					p.Status.Phase = corev1.PodRunning
+				}
+			})
+
+			It("should return success", func() {
+				Eventually(waitCh, defaults.WaitDurationResources).Should(BeClosed())
+			})
+		})
+
+		Context("and are eventually running", func() {
+			It("should return success", func() {
+				go func() {
+					time.Sleep(defaults.WaitRetryPeriod * 2)
+					for _, p := range initialPods {
+						p.Status.Phase = corev1.PodRunning
+						_ = client.Update(context.TODO(), p)
+					}
+				}()
+
+				Eventually(waitCh, defaults.WaitDurationResources).Should(BeClosed())
+			})
+		})
+	})
+
+	When("the Pod does not initially exist but is eventually created and running", func() {
+		It("should return success", func() {
+			go func() {
+				time.Sleep(defaults.WaitRetryPeriod * 2)
+				_ = client.Create(context.TODO(), newPod("test-Pod1", corev1.PodRunning))
+			}()
+
+			Eventually(waitCh, defaults.WaitDurationResources).Should(BeClosed())
+		})
+	})
+
+	When("retrieval of the running Pod list initially fails but eventually succeeds", func() {
+		BeforeEach(func() {
+			initialPods = append(initialPods, newPod("test-Pod", corev1.PodRunning))
+			listError = errors.New("mock error")
+		})
+
+		It("should return success", func() {
+			Eventually(waitCh, defaults.WaitDurationResources).Should(BeClosed())
+		})
+	})
+
+	When("the Pod never becomes running", func() {
+		BeforeEach(func() {
+			defaults.WaitDurationResources = 300 * time.Millisecond
+			defaults.WaitRetryPeriod = 50 * time.Millisecond
```

Comment:
Won't this override these "defaults" for any test that would run after these tests?

Reply:
Yeah what I meant is it will have effect even after these tests are done running since you don't reset them to their "original values" and these are globals, so that might interfere with other tests that use these values.
Whether that's good or bad I can't tell, but not sure it's what you intended to ahieve.

Reply:
yeah I thought about that but really every other UT that uses these functions will have to set these values appropriately (I already did the same in others). But I can add BeforeSuite/After Suite to be complete.

Reply:
Thanks!